mass/core/optimal_filtering.py

```diff
@@ -296,6 +299,126 @@ def report(self, std_energy: float = 5898.8):
+            A 1-d array, a single pulse record, or a 2-d array, each row a pulse records.
+
+        Returns
+        -------
+        tuple[np.ndarray, np.ndarray]
+            1. The optimally filtered value, or an array (one per row) if the input is a 2-d array.
+            2. The phase, or arrival-time estimate in samples. Same shape as the filtered value.
+
+        Raises
+        ------
+        AssertionError
+            If the input array is the wrong length
+        """
+        x = np.asarray(x)
+        if x.ndim == 1:
+            x = x.reshape((1, len(x)))
+        nrec, nsamp = x.shape
+        assert nsamp == len(self.values) + self.convolution_lags - 1
+        conv = np.zeros((5, nrec), dtype=float)
+        conv[0, :] = np.dot(x[:, 0:-4], self.values)
+        conv[1, :] = np.dot(x[:, 1:-3], self.values)
+        conv[2, :] = np.dot(x[:, 2:-2], self.values)
+        conv[3, :] = np.dot(x[:, 3:-1], self.values)
+        conv[4, :] = np.dot(x[:, 4:], self.values)
+
+        # Least-squares fit of 5 values to a parabola.
+        # Order is row 0 = constant ... row 2 = quadratic coefficients.
+        param = np.dot(self.FIVELAG_FITTER, conv)
+        peak_x = -0.5 * param[1, :] / param[2, :]
+        peak_y = param[0, :] - 0.25 * param[1, :]**2 / param[2, :]
+        return peak_y, peak_x
+
+
+@dataclass(frozen=True)
+class FilterATS(Filter):
+    """Represent an optimal filter according to the arrival-time-safe, single-lag design of 2015.
+
+    Returns
+    -------
+    FilterATS
+        An optimal filter, for convolution with data (at 5 lags, obviously)
+    """
+
+    def __post_init__(self):
+        assert self.convolution_lags == 1
+        assert self.dt_values is not None
+
+    @property
+    def is_arrival_time_safe(self):
+        """Is this an arrival-time-safe filter?"""
+        return True
+
+    @property
+    def _filter_type(self):
+        return "ats"
+
+    def filter_records(self, x: npt.ArrayLike) -> tuple[np.ndarray, np.ndarray]:
+        """Filter one microcalorimeter record or an array of records.
+
+        Parameters
+        ----------
+        x : npt.ArrayLike
+            A 1-d array, a single pulse record, or a 2-d array, each row a pulse records.
+
+        Returns
+        -------
+        tuple[np.ndarray, np.ndarray]
+            1. The optimally filtered value, or an array (one per row) if the input is a 2-d array.
+            2. The phase, or arrival-time estimate in samples. Same shape as the filtered value.
+
+        Raises
+        ------
+        AssertionError
+            If the input array is the wrong length
+        """
+        x = np.asarray(x)
+        if x.ndim == 1:
+            x = x.reshape((1, len(x)))
+        _, nsamp = x.shape
+        print("Heyho! ", nsamp, x.shape, len(self.values), type(self))
+
+        assert nsamp == len(self.values)
+        conv0 = np.dot(x, self.values)
+        conv1 = np.dot(x, self.dt_values)
+        ArrivalTime = conv1 / conv0
```

Comment:
Capital letters conventionally refer to a class, rename `ArrivalTime` to `arrival_time` or `filt_phase` or similar. Github even shows it as orange, just like the class name `FilterMaker` and others.

Reply:
Good point. This is _very_ old code. I guess this is relic of an earlier time.